### PR TITLE
[Feature] allow generated repo_configs in the `gazelle` rule

### PR DESCRIPTION
**Pull request by grahamjenson**
TLDR: I want to be able to pass the `gazelle` rule a generated repo_config so that I don't have to manually update dependencies from `go.mod`

Using the `update-repos` workflow to update repositories from a go.mod is pretty painful. To simplify this I want to be able to generate a repository config file automatically and have gazelle use it, without having to run `update-repos`. To do this I wrote a repository rule:

```
def _go_mod_impl(ctx):
    ctx.file("BUILD", content = 'exports_files(["repositories.bzl"])')
    ctx.file("WORKSPACE", content = "workspace(name = \"%s\")" % ctx.name)

    bzl = 'load("@bazel_gazelle//:deps.bzl", "go_repository")\n\ndef go_repositories():\n    pass\n\n# gazelle:repository_macro repositories.bzl%go_repositories'

    ctx.file("repositories.bzl", content = bzl)

    _gazelle = "@bazel_gazelle_go_repository_tools//:bin/gazelle"
    gazelle = ctx.path(Label(_gazelle))

    result = ctx.execute(
        [
            gazelle,
            "update-repos",
            "-from_file=%s" % ctx.path(ctx.attr.go_mod),
            "-to_macro=repositories.bzl%go_repositories",
        ],
    )

    if result.return_code:
        fail("failed go_mod %s: %s" % (ctx.name, result.stderr))
    if result.stderr:
        print("go_mod: " + result.stderr)


go_mod_download = repository_rule(
    implementation = _go_mod_impl,
    attrs = {
        "go_mod": attr.label(
            mandatory = True,
            allow_single_file = True,
        ),
    },
)
```

That is called like:

```
go_mod_download(
    name = "go_modules",
    go_mod = "//:go.mod",
)
load("@go_modules//:repositories.bzl", default_go_repositories = "go_repositories")
default_go_repositories()
```

This works great. Except because of the way gazelle optimizes resolution of repositories, it has made running gazelle **very slow**.

To solve this I want to be able to pass my generated `@go_modules//:repositories.bzl` as a repo config to gazelle, to tell it about all the repositories I already have defined. For example:

```
gazelle(
    name = "gazelle",
    repo_config = "@go_modules//:repositories.bzl",
)
```

This PR lets gazelle work with a generated repo_config. This is also very extensible, because now I can concat a bunch of files that have `go_repositories` in them, into a single repository config and have gazelle know what is what. 

With this approach, I can run `go get -u <package>` to update a package then run `bazel build <rule>` and it will have the new package immediately without having to ru `update-repos` AND if I run `bazel run gazelle` it doesn't take minutes.

### Other approach

I tried using something like  `# gazelle:repository_macro bazel-repo/external/repositories.bzl%go_repositories` this doesn't work because when running `bazel run gazelle` the file might not exist yet. So gazelle rule must know about this file before hand anyway.

### CODE REVIEW

**Comment by jayconrod** [comment]
Please open an issue to discuss this. I'd prefer not to review code until we've agreed on a design.

You may want to read through #631 first. Gazelle won't execute Starlark code, and it won't invoke Bazel, which means that Gazelle can't directly access configuration files in other workspaces. It's okay for rules and repository rules that invoke Gazelle to depend on files (possibly generated files) in other workspaces though and pass those in.

Also, take a look at the `gazelle_dependencies` function. It accepts a `config` parameter that defaults to `@//:WORKSPACE`, but could be set to a generated file in another repository. I think that was added by the Uber folks, who were doing something similar. If that works for you, then maybe it just need to be better documented.

**Comment by grahamjenson** [comment]
Hey, 
I tried running my above solution with `go_repository_default_config = "@go_modules//:repositories.bzl"` and `gazelle_dependencies(go_repository_default_config = "@go_modules//:WORKSPACE")` with a variety of directives. It looks like that argument `go_repository_default_config` can take a file where the directive `# gazelle:repository_macro ` is a label which is super confusing, because I am unable to use a label and point to the function `go_repositories` which is needed. 

The core issue with this approach is that running `bazel run gazelle` is very slow compared because it is not be able create a list of known imports as the repository config at https://github.com/bazelbuild/bazel-gazelle/blob/960ddc5b3e5c4a7da610b27b7d122f40ec56a188/cmd/gazelle/fix-update.go#L131 always resolves to <my_repo>/WORKSPACE and not the workspace set in the `gazelle_dependencies` function. I am not sure if there is some trick to fi this, maybe make `@go_modeules` look like a pseudo go_repository or something?

I have created https://github.com/bazelbuild/bazel-gazelle/issues/752 to have further discussion.

**Comment by robfig** [comment]
Cleaning up old PRs.. this one seems to have stalled, and the final design seems likely to be not related to this, so I'll close it for now. Thanks
